### DISCUSSION TITLE
example howto getOrders

### DISCUSSION THREAD

**Discussion by humpataa** [amzn/selling-partner-api-models#3140] [General]
Hi,

I am making my first steps with this API and have successfully retrieved datas via getOrder.
However, I always get a 400 when I try to request the list of orders ... can someone please post a working example?

This is working fine:
$result = $api->getOrder("302-9972095-7840xxx");

This does not:
$result = $api->getOrders(array('A1PA6795UKMFR9'), '2021-09-01T22:00:00Z');

Any help is greatly appreciated!
Cheers!

**Comment by NetRushJacob** [comment]
More information is likely needed to answer this. 
What's in 'getOrders' method?

**Reply by humpataa** [reply]
Hey Jacob, it's the API function:
https://github.com/amzn/selling-partner-api-docs/blob/main/references/orders-api/ordersV0.md#getorders

**Reply by NetRushJacob** [reply]
I get that's what the method points to, but what is the actual code in getOrders? 
From this line: getOrders(array('...'), '2021-09-01T22:00:00Z'); 
We (I?) cannot determine what's actually being called. 
Is that date pointing to LastUpdatedAfter, LastUpdatedBefore, CreatedAfter, or CreatedBefore?
Is it doing a post instead of a get?
Does it possibly have a miss-spelling in a parameter or otherwise malformed? 
Was the contents of getOrders(...) method generated from the Swagger or manually created?

**Reply by humpataa** [reply]
Well - I don't know, I am using the API, I am not developing it. 
I have tried calling getOrders with all possible kind of arguments (referring to the docs) but it always returns an error message (invalid arguments). Here is another test which fails:

```
$marketplace_ids = array('A1PA6795UKMFR9'); 
$created_after = '2021-09-15T22:00:00Z'; 
$created_before = ''; 
$last_updated_after = ''; 
$last_updated_before = ''; 
$order_statuses = array(''); 
$fulfillment_channels = array(''); 
$payment_methods = array(''); 
$buyer_email = '';
$seller_order_id = ''; 
$max_results_per_page = 50; 
$easy_ship_shipment_statuses = array(''); 
$next_token = ''; 
$amazon_order_ids = array(''); 
$actual_fulfillment_supply_source_id = ''; 
$is_ispu = False; 
$store_chain_store_id = ''; 

$result = $api->getOrders($created_after, $created_before, $last_updated_after, $last_updated_before, $order_statuses, $marketplace_ids, $fulfillment_channels, $payment_methods, $buyer_email, $seller_order_id, $max_results_per_page, $easy_ship_shipment_statuses, $next_token, $amazon_order_ids, $actual_fulfillment_supply_source_id, $is_ispu, $store_chain_store_id);

```

So my question is: how to call this function properly?
Thank you!

**Reply by NetRushJacob** [reply]
I wonder if I'm missing something or if we have a terminology translation issue. 
Is this client generated from the swagger? 
A downloaded client package? 
What language is it in?

For reference / comparison, this is our hand-coded 'GetOrders' C# method (GetOrdersRequest matches the model in the swagger):
`        public async Task<GetOrdersResponse> GetOrdersAsync(GetOrdersRequest getOrdersRequest)
        {
            if (getOrdersRequest.CreatedAfter.HasValue && getOrdersRequest.LastUpdatedAfter.HasValue)
            {
                throw new Exception("Cannot create a GetOrdersRequest with both CreatedAfter and LastUpdatedAfter specified");
            }

            var client = GetClient();

            var request = new RestRequest("orders/v0/orders", Method.GET)
                    .WithQueryParameterIfNotNull("NextToken", getOrdersRequest.NextToken)
                    .WithQueryParameterIfNotNull("MarketplaceIds", getOrdersRequest.MarketplaceIds.ToArrayQueryParameter());
            if (string.IsNullOrEmpty(getOrdersRequest.NextToken))
            {
                request = request
                    .WithQueryParameterIfNotNull("CreatedAfter", getOrdersRequest.CreatedAfter?.ToString("yyyy-MM-ddTHH:mm:ssZ"))
                    .WithQueryParameterIfNotNull("CreatedBefore", getOrdersRequest.CreatedBefore?.ToString("yyyy-MM-ddTHH:mm:ssZ"))
                    .WithQueryParameterIfNotNull("LastUpdatedAfter", getOrdersRequest.LastUpdatedAfter?.ToString("yyyy-MM-ddTHH:mm:ssZ"))
                    .WithQueryParameterIfNotNull("LastUpdatedBefore", getOrdersRequest.LastUpdatedBefore?.ToString("yyyy-MM-ddTHH:mm:ssZ"))
                    .WithQueryParameterIfNotNull("OrderStatuses", getOrdersRequest.OrderStatuses.ToArrayQueryParameter())
                    //.WithQueryParameterIfNotNull("MarketplaceIds", getOrdersRequest.MarketplaceIds.ToArrayQueryParameter())
                    .WithQueryParameterIfNotNull("FulfillmentChannels", getOrdersRequest.FulfillmentChannels.ToArrayQueryParameter())
                    .WithQueryParameterIfNotNull("PaymentMethods", getOrdersRequest.PaymentMethods.ToArrayQueryParameter())
                    .WithQueryParameterIfNotNull("BuyerEmail", getOrdersRequest.BuyerEmail)
                    .WithQueryParameterIfNotNull("SellerOrderId", getOrdersRequest.SellerOrderId)
                    .WithQueryParameterIfNotNull("MaxResultsPerPage", getOrdersRequest.MaxResultsPerPage?.ToString())
                    .WithQueryParameterIfNotNull("EasyShipShipmentStatuses", getOrdersRequest.EasyShipShipmentStatuses.ToArrayQueryParameter())
                    .WithQueryParameterIfNotNull("AmazonOrderIds", getOrdersRequest.AmazonOrderIds.ToArrayQueryParameter());
            }

            var response = await client.ExecuteAsync<GetOrdersResponse>(SignedSellerRequest(request, client));
            response?.Data?.SetHeaders(response.Headers);
            return response?.Data;
        }`

**Reply by humpataa** [reply]
Thank you - well, you are right, I forgot to mention - I am using this PHP code: https://github.com/jlevers/selling-partner-api
The exception I get is this one:

```
Exception [400] {
  "errors": [
    {
      "code": "InvalidInput",
      "message": "Application do not have access to some or all requested resource",
      "details": ""
    }
```

But as I wrote in my first post - calling getOrder with a specific orderId as parameter works as expected!
Does a request to getOrders need special permissions? Am I missing something? Any idea?

**Reply by NetRushJacob** [reply]
GetOrder and GetOrders should use the same permissions, to my knowledge. 
A quick look at the client library, everything looks normal to me.

That error indicates that your application might not have access...maybe to PII? I would expect those nodes to just be blank, not error out, if that were the case.

What permissions are you requesting in your app?

**Reply by humpataa** [reply]
What do you mean with "to PII"? Nodes? It's my seller-central, my app, my developer account, nothing is shared, nothing open.

Here's what I have enabled in roles setting for the app:
Product Listing, Pricing, Amazon Fulfilment, Buyer Com, Buyer Sol, Selling Partner Insights, Finance and Accunting, Inventory and Order Tracking

I don't know but the app has "status" draft and whenever I authorize it, the refresh token changes, but my PHP works with the old token and the new!? So what is this for anyway? There doesn't seem to be another "status" either. Personally I find this to be security overkill: keys, ids, tokens, secret keys, identifiers, roles. What the hell. After trying around for weeks now, I am quite annoyed. I just want a list of our orders, I am not building an atomic bomb.

**Reply by NetRushJacob** [reply]
PII: Personally Identifiable Information
PII is (or will be) restricted behind further security beyond the standard authentication/authorization. This is likely partly due to GDPR and CCPA and partly due to actually caring about their customers' data and privacy.

Nodes: Nearly every single API is requested and responding with JSON. JSON is comprised of nodes, or containers of data.  Related to Orders, there are certain nodes of the response (like the BuyerInfo node) that will not return if the application requesting is not authorized to access PII, or not submitting the Restricted Data Token.

'Draft' just means it hasn't been reviewed and approved by Amazon and won't be available on the partner network (or, if it's been published before, the latest changes won't display).

The refresh tokens likely have a short life after re-authorizing/refreshing, but I'm not an expert there, so don't quote me on that.
I find Amazon's focus on security, especially in the realm of PII, refreshing and on target to be an ideal standard for other companies to aspire to. 
These APIs are geared for creating applications _for interacting with others,_ and have to cover a lot of use cases with this security. 

All of the above being said, the error you posted implies that your application does not have access to all or part of the GetOrders API. To me, that implies the specific parts of the model that could be restricted beyond the basic call, like PII.  **Why it's erroring instead of just not returning a response without those JSON nodes, I'm not sure**.

See: https://github.com/amzn/selling-partner-api-docs/discussions?discussions_q=RDT for discussions around the Restricted Data Token (RDT)

Another possibility is that the client you are using is somehow broken, although the error implies permissions more than an actual broken call.

**Reply by humpataa** [reply]
Thank you for the details Jacob, I didn't mean to be rude, I am just a little tired and will most likely just leave it and stick to do everything by hand as we have done for months and years now.

**Reply by NetRushJacob** [reply]
Text can sound rude; I try to counter that with answering each point or concern.  Reaching out to support directly might lead to some quick and easy answers that I couldn't identify.

**Reply by humpataa** [reply]
Well - how do I reach out directly? Thought this board is supposed to be the "support"?

**Reply by NetRushJacob** [reply]
I think you actually just create a case in Seller Central; maybe mention in the case that you are developing an app (as apposed to trying to sell, even if you're developing an app only for yourself).

Guidelines: https://sellercentral.amazon.com/gp/help/external/GK3V2JYCN2P28BFX?language=en_US&ref=efph_GK3V2JYCN2P28BFX_cont_69036

There used to be an email way back when this was all starting out (before it was 'public' but I'm failing to find it now.  SellerCentral support case is probably the most direct route, now.

**Reply by maurijph** [reply]
> Hello does any know how to get PII Approval? I submitted that questionnaire 3 times and they rejected each time and sent me this response.
> 
> AUP - Acceptable Use Policy
> 
> https://sellercentral.amazon.com/mws/static/policy?documentType=AUP&locale=us_US
> 
> Data usage 4.1. Question - Describe why you require Personally Identifiable Information to build your application or feature.
> 
> RDA - Restricted Data Access
> 
> https://sellercentral.amazon.com/mws/static/policy?documentType=DPP&locale=us_US
> 
> Network Protection 1.1 Question - Describe the network protection controls used by your organization to restrict public access to databases, file servers, and desktop/developer endpoints.
> 
> Access Management 1.2 and Least Privilege Principle 1.3. Question - Describe how your organization individually identifies employees who have access to Amazon information, and restricts employee access to Amazon information on a need-to-know basis.
> 
> Asset Management 2.3. Question - Describe the mechanism your organization has in place to monitor and prevent Amazon Information from being accessed from employee personal devices (such as USB flash drives, cellphones) and how are you alerted in the event such incidents occur.
> 
> Data Governance 2.2 Question - Provide your organization´s privacy and data handling policies to describe how Amazon data is collected, processed, stored, used shared and disposed. You may provide this in the form of a public website URL.
> 
> Encryption at Rest 2.4 Question - Describe where your organization stores Amazon Information at rest and provide details on any encryption algorithm used.
> 
> Data Retention 2.1 Question - Describe how your organization backups or archives Amazon Information and provide details on any encryption algorithm used.
> 
> Logging and Monitoring 2.6 Question - Describe how your organization monitors, detects, and logs malicious activity in your application(s).
> 
> Secure Coding Practices 2.5 Question - How is Personally Identifiable Information (PII) protected during testing?
> 
> Secure Coding Practices 2.5 Question - What measures are taken to prevent exposure of credentials?
> 
> Vulnerability Management 2.7 Question - How do you track remediation progress of findings identified from vulnerability scans and penetration tests?
> 
> Vulnerability Management 2.7 Question - How do you address code vulnerabilities identified in the development lifecycle and during runtime?
> 
> Vulnerability Management 2.7 Question - Who is responsible for change management and how is their access granted? Please specify job title.
> 
> No changes have been made to your SP-API access at this time.
Hi @Sohailayubbutt 
did you solve these questions? I have the exact same issue.
Let me know how you answered
Regards

**Comment by gjaon** [comment]
const orders = async (req, res) => {
    try {
        const createdAfter = moment().subtract(30, 'days').toISOString();
        const req_params = {
            "MarketplaceIds": [marketplace_id],
            "createdAfter": createdAfter
        };

        const authTokens = await authenticate();

        const queryString = new URLSearchParams(req_params).toString();
        const url = `${endpoint}/orders/v0/orders?${queryString}`;

        const response = await axios.get(url, {
            "headers": {
                "x-amz-access-token": authTokens.access_token,
                "Content-Type": "application/json"
            }
        });

        res.status(200).json(response.data);
    } catch (error) {
        console.error(error.response || error);
        res.status(500).json({ message: "Error getting orders", error: error });
    }
}; 
This is what I wrote, and i get the error of 400. meanwhile I am able to get the access_token with a different call, so I am passing the access token here. can anyone help?

**Reply by gjaon** [reply]
I get this same response for the sandbox also

**Reply by NetRushJacob** [reply]
400 is bad request. Things I would look at compared to my C# implementation would be:

1. 'createdAfter' param name vs my 'CreatedAfter' (shouldn't matter, but is a difference)
2. toISOString() would return YYYY-MM-DDTHH:mm:ss.sssZ compared to my formatting yyyy-MM-ddTHH:mm:ssZ   
3. May not be relevant, but Orders might also be looking for a [Restricted Data Token](https://developer-docs.amazon.com/sp-api/docs/authorization-with-the-restricted-data-token) instead of the regular auth token

General suggestion when dealing with the Orders API: Use LastUpdatedAfter over CreatedAfter.  Reasoning: Amazon loves to make changes on orders in the past, even longer ago than 30 days. This is especially true of the seller makes use of Subscribe and Save, which appears to create the 'next' order when the 'current' order ships. That 'next' order has Purchase Date on the date it was created, but won't charge and ship until it's scheduled to, likely month(s) in the future.

**Reply by gjaon** [reply]
I believe you are correct Jacob, I used this endpoint `https://sellingpartnerapi-eu.amazon.com/orders/v0/orders?MarketplaceIds=ATVPDKIKX0DER&CreatedAfter=2020-10-10&MaxResultPerPage=2` which I got from the docs so I had to make the variables dynamic by doing `${endpoint}/orders/v0/orders?MarketplaceIds=${marketplace_id}&CreatedAfter=${createdAfter}&MaxResultPerPage=2` 

It just shows that my implementation did not parse the parameters accurately
